### PR TITLE
chore(repo): migrate to eslint

### PR DESCRIPTION
**Pull request by HuiSF**
#### Description of changes

1. Use `typescript` 5.0.2 in the workspace!
2. Unified `tsconfig.json` setup and its branching across packages
    * ts will check unit tests too to ensure we are using correct type to ensure unit tests are accurate
    * virtual paths are enabled in each package, no more `../../../module` import paths
4. Migrated to eslint and remove tslint
5. Ran `eslint --fix` for each package to auto fix lint issues (approximately 80% of the changes), plus manual fixes (approximately 20% of the changes)

NOTE: This PR upgrades typescript 5.0.2 which breaks the existing jest set up, this PR will depends on https://github.com/aws-amplify/amplify-js/pull/12607 to work (hence marked as Draft PR, and currently `__tests__` is ignored by eslint ).

NOTE: Please **DO NOT** click on update branch button. Thanks

### What's new after rebasing off origin/main

1. [chore(repo): update jest setup script to work with eslint and tsconfig](https://github.com/aws-amplify/amplify-js/pull/12620/commits/4a70b058d6e6f06c788ce0f829a93488d971bbec)
2. necessary changes made to `__tests__` to let the unit tests run (in the commits that are with message: "squash to the previous"...)

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/awsClients/cognitoIdentity/types.ts [hunk 1]
```diff
@@ -32,8 +32,8 @@ export interface Credentials {
  *
  * The input for {@link GetCredentialsForIdentityCommand}.
```
Comment (cshfang) [inline packages/core/src/awsClients/cognitoIdentity/types.ts:33]: These links are emitting warnings
> The type 'GetCredentialsForIdentityCommand' is undefined

> The type 'GetIdCommand' is undefined

File: packages/core/src/awsClients/pinpoint/types.ts [hunk 2]
```diff
@@ -377,7 +376,7 @@ export interface EventsResponse {
  *
  * The input for {@link GetInAppMessagesCommand}.
```
Comment (cshfang) [inline packages/core/src/awsClients/pinpoint/types.ts:377]: Ditto these links
> The type 'GetInAppMessagesCommand' is undefined

File: packages/core/src/utils/WordArray.ts [hunk 3]
```diff
@@ -12,8 +12,8 @@ import { cryptoSecureRandomInt } from './cryptoSecureRandomInt';
  */
 function hexStringify(wordArray: WordArray): string {
 	// Shortcuts
-	const words = wordArray.words;
-	const sigBytes = wordArray.sigBytes;
+	const {words} = wordArray;
+	const {sigBytes} = wordArray;
```
Comment (cshfang) [inline packages/core/src/utils/WordArray.ts:16]: Nit: Padding spaces inside curly braces don't applied to destructuring?

File: .prettierrc.js [hunk 4]
```diff
@@ -1,5 +1,6 @@
 module.exports = {
-	trailingComma: 'es5',
+	trailingComma: 'all',
+	printWidth: 80,
```
[the remark this answers is not in the document]
Comment (HuiSF) [reply .prettierrc.js:3]: Without specifying the `printWidth` here, the auto format on save will not work correctly mixing with `max-len`.

File: packages/core/src/awsClients/cognitoIdentity/types.ts (hunk 1 above)
Comment (HuiSF) [reply packages/core/src/awsClients/cognitoIdentity/types.ts:33]: Yes, meant to check with @AllanZhengYP about this it seems that these docs are copied over from SDK code, needed a clean up with a correct way.

File: packages/core/src/utils/WordArray.ts (hunk 3 above)
Comment (HuiSF) [reply packages/core/src/utils/WordArray.ts:16]: Oh, good catch let me check the rules.

File: packages/core/src/awsClients/cognitoIdentity/types.ts (hunk 1 above)
Comment (AllanZhengYP) [reply packages/core/src/awsClients/cognitoIdentity/types.ts:33]: Yes they are copied over. For this PR can we disable such warnings for `package/*/src/awsClients/*/types.ts`? The copy-over script is from [dts-bundler](https://github.com/aws-amplify/amplify-js/tree/main/scripts/dts-bundler). In longer term we can remove the comments with pattern`\*The {input,output} of \{@link .*Command\}.` in this script.

Comment (HuiSF) [reply packages/core/src/awsClients/cognitoIdentity/types.ts:33]: Currently these are warnings that are not blocking. Should we leave it as is as a reminder that we will need to do about it? wdyt? @cshfang @AllanZhengYP

File: packages/storage/package.json [hunk 5]
```diff
@@ -105,8 +106,7 @@
 		"@aws-amplify/react-native": "1.0.5",
 		"@rollup/plugin-json": "6.0.1",
 		"@rollup/plugin-typescript": "11.1.5",
-		"rollup": "3.29.4",
-		"typescript": "5.0.2"
```
Comment (kvramyasri7) [inline packages/storage/package.json:109]: Question: any reason removing this `typescript`?

File: packages/storage/src/providers/s3/apis/internal/list.ts [hunk 6]
```diff
@@ -5,32 +5,32 @@ import { AmplifyClassV6 } from '@aws-amplify/core';
...
-import { logger } from '../../../../utils';
+	listObjectsV2,
+} from '~/providers/s3/utils/client';
+import { getStorageUserAgentValue } from '~/providers/s3/utils/userAgent';
+import { logger } from '~/utils';
 
 const MAX_PAGE_SIZE = 1000;
 
-type ListInputArgs = {
+interface ListInputArgs {
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/apis/internal/list.ts:25]: Question: Any reason why `type` > `interface` here and elsewhere? Is there a decision already taken which I am not aware of?

File: packages/storage/src/providers/s3/utils/client/runtime/xmlParser/dom.ts [hunk 7]
```diff
@@ -31,22 +32,20 @@ const parseXmlNode = (node: Node): any => {
...
 			if (!isNamespaceAttributeName(attr.nodeName)) {
 				nodeValue[attr.nodeName] = attr.nodeValue!;
 			}
 		}
 
 		// convert child nodes
 		if (node.children.length > 0) {
-			for (let i = 0; i < node.children.length; i++) {
-				const child = node.children[i];
+			for (const child of node.children) {
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/utils/client/runtime/xmlParser/dom.ts:48]: 🤩

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 8]
```diff
@@ -2,25 +2,26 @@
...
-type S3ApiOptions = {
+interface S3ApiOptions {
 	accessLevel?: StorageAccessLevel;
 	targetIdentityId?: string;
 	useAccelerateEndpoint?: boolean;
-};
+}
 
-type ResolvedS3ConfigAndInput = {
+export interface ResolvedS3ConfigAndInput {
```
Comment (kvramyasri7) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:19]: Any reason in why exporting this one?

Comment (HuiSF) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:19]: ts inline doc requires it.

File: packages/storage/package.json (hunk 5 above)
Comment (HuiSF) [reply packages/storage/package.json:109]: This dev dep is now a workspace dev dep.

File: packages/storage/src/providers/s3/apis/internal/list.ts (hunk 6 above)
Comment (HuiSF) [reply packages/storage/src/providers/s3/apis/internal/list.ts:25]: https://www.typescriptlang.org/docs/handbook/advanced-types.html#interfaces-vs-type-aliases

File: package.json [hunk 9]
```diff
@@ -37,6 +38,9 @@
 		"ts-coverage": "lerna run ts-coverage",
 		"prepare": "./scripts/set-preid-versions.sh"
 	},
+	"husky": {
+		"hooks": {}
+	},
```
Comment (cshfang) [inline package.json:43]: I think we can remove this

**Review by cshfang — APPROVED** [review]

**Review by cshfang — APPROVED** [review]

**Review by jimblanc — APPROVED** [review]
Thank you for taking this on, looks good to me. Paid special attention to analytics, core, and umbrella package. Have we run the integ test suite against this to double check? Also looks like we have a minor bundle size regression that needs to be adjusted.

**Review by AllanZhengYP — APPROVED** [review]

**Pull request closed without merging**